src/main/java/com/linecorp/armeria/common/ServiceInvocationContext.java

```diff
@@ -149,6 +154,116 @@ public final EventLoop eventLoop() {
+     */
+    public final EventExecutor contextAwareEventExecutor() {
+        return new ServiceInvocationContextAwareEventExecutor(this, eventLoop());
+    }
+
+    /**
+     * Returns a {@link Executor} that will execute callbacks in the given
+     * {@code executor}, making sure to propagate the current invocation context
+     * into the callback execution. It is generally preferred to use
+     * {@link #contextAwareEventExecutor()} to ensure the callback stays on the
+     * same thread as well.
+     */
+    public final Executor makeContextAware(Executor executor) {
+        return runnable -> executor.execute(makeContextAware(runnable));
+    }
+
+    /**
+     * Returns a {@link Callable} that makes sure the current invocation context
+     * is set and then invokes the input {@code callable}.
+     */
+    public final <T> Callable<T> makeContextAware(Callable<T> callable) {
+        ServiceInvocationContext propagatedContext = this;
+        return () -> {
+            boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+            try {
+                return callable.call();
+            } finally {
+                if (mustResetContext) {
+                    removeCurrent();
+                }
+            }
+        };
+    }
+
+    /**
+     * Returns a {@link Runnable} that makes sure the current invocation context
+     * is set and then invokes the input {@code runnable}.
+     */
+    public final Runnable makeContextAware(Runnable runnable) {
+        ServiceInvocationContext propagatedContext = this;
+        return () -> {
+            boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+            try {
+                runnable.run();
+            } finally {
+                if (mustResetContext) {
+                    removeCurrent();
+                }
+            }
+        };
+    }
+
+    /**
+     * Returns a {@link GenericFutureListener} that makes sure the current invocation
+     * context is set and then invokes the input {@code listener}.
+     */
+    public final <T extends Future<?>> GenericFutureListener<T> makeContextAware(
+            GenericFutureListener<T> listener) {
+        ServiceInvocationContext propagatedContext = this;
+        return new GenericFutureListener<T>() {
+            @Override
+            public void operationComplete(T future) throws Exception {
+                boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+                try {
+                    listener.operationComplete(future);
+                } finally {
+                    if (mustResetContext) {
+                        removeCurrent();
+                    }
+                }
+            }
+
+            // Delegate equality checks to the original listener so removeListener works.
+            @Override
+            public boolean equals(Object obj) {
+                return listener.equals(obj);
```

Comment:
newSucceededFuture and newFailedFuture do not have that problem because removeListener(s) is no-op.

However, newPromise() and newProgressivePromise() do have the problem. Perhaps I could make necessary changes in Netty? :-)